cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteProgramRunDispatcher.java

```diff
@@ -0,0 +1,149 @@
+ */
+public class RemoteProgramRunDispatcher implements ProgramRunDispatcher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteProgramRunDispatcher.class);
+
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(
+          new GsonBuilder())
+      .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+      .registerTypeAdapter(ApplicationClass.class, new ApplicationClassCodec())
+      .registerTypeAdapter(Requirements.class, new RequirementsCodec())
+      .registerTypeAdapter(RunId.class, new RunIds.RunIdCodec())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+      .create();
+  private final Store store;
+  private final ProgramRunnerFactory programRunnerFactory;
+  private final RemoteTaskExecutor remoteTaskExecutor;
+  private ProgramRunnerFactory remoteProgramRunnerFactory;
+  private TwillRunnerService remoteTwillRunnerService;
+
+  @Inject
+  public RemoteProgramRunDispatcher(CConfiguration cConf,
+      MetricsCollectionService metricsCollectionService,
+      RemoteClientFactory remoteClientFactory, Store store,
+      ProgramRunnerFactory programRunnerFactory) {
+    this.store = store;
+    this.programRunnerFactory = programRunnerFactory;
+    this.remoteTaskExecutor = new RemoteTaskExecutor(cConf, metricsCollectionService,
+        remoteClientFactory, RemoteTaskExecutor.Type.SYSTEM_WORKER);
+  }
+
+  /**
+   * Optional guice injection for the {@link ProgramRunnerFactory} used for remote execution. It is
+   * optional because in unit-test we don't have need for that.
+   */
+  @Inject(optional = true)
+  void setRemoteProgramRunnerFactory(
+      @Constants.AppFabric.RemoteExecution ProgramRunnerFactory runnerFactory) {
+    this.remoteProgramRunnerFactory = runnerFactory;
+  }
+
+  /**
+   * Optional guice injection for the {@link TwillRunnerService} used for remote execution. It is
+   * optional because in unit-test we don't have need for that.
+   */
+  @Inject(optional = true)
+  void setRemoteTwillRunnerService(
+      @Constants.AppFabric.RemoteExecution TwillRunnerService twillRunnerService) {
+    this.remoteTwillRunnerService = twillRunnerService;
+  }
+
+  @Override
+  public ProgramController dispatchProgram(ProgramRunInfo programRunInfo) throws Exception {
+    RunId runId = programRunInfo.getRunId();
+    LOG.debug("Start Program Run operation for Run ID: {}", runId.getId());
+    RunnableTaskRequest request = RunnableTaskRequest.getBuilder(
+        ProgramRunDispatcherTask.class.getName()).withParam(GSON.toJson(programRunInfo)).build();
+    byte[] result = remoteTaskExecutor.runTask(request);
+    RemoteProgramRunDispatcherResponse response = GSON.fromJson(
+        new String(result, StandardCharsets.UTF_8), RemoteProgramRunDispatcherResponse.class);
+    if (!response.isProgramRunSuccessful()) {
+      throw new Exception("Program Run Operation Failed.");
```

Comment:
please add more contextual details here.

Reply:
Added comment. This is an unexpected exception. Response should always be true. Otherwise, http callee with throw an exception which be handled by http exception handler at the caller side.